service/storage_proxy.cc

```diff
@@ -1177,10 +1177,44 @@ future<bool> paxos_response_handler::accept_proposal(const paxos::proposal& prop
 future<> paxos_response_handler::learn_decision(paxos::proposal decision, bool allow_hints) {
     tracing::trace(tr_state, "learn_decision: committing {} with cl={}", decision, _cl_for_learn);
     paxos::paxos_state::logger.trace("CAS[{}] learn_decision: committing {} with cl={}", _id, decision, _cl_for_learn);
-    std::array<std::tuple<paxos::proposal, schema_ptr, dht::token>, 1> m{std::make_tuple(std::move(decision), _schema, _key.token())};
     // FIXME: allow_hints is ignored. Consider if we should follow it and remove if not.
     // Right now we do not store hints for when committing decisions.
-    return _proxy->mutate_internal(std::move(m), _cl_for_learn, false, tr_state, _permit, _timeout);
+    future<> f_cdc = make_ready_future<>();
+    if (_schema->cdc_options().enabled()) {
+        auto update_mut = decision.update.unfreeze(_schema);
+        const auto base_tbl_id = update_mut.column_family_id();
+        std::vector<mutation> update_mut_vec;
+        update_mut_vec.push_back(std::move(update_mut));
+
+        if (_proxy->get_cdc_service() && _proxy->get_cdc_service()->needs_cdc_augmentation(update_mut_vec)) {
+            // CDC mutations are sent down another route
```

Comment:
I have hard time figuring out from this comment what's going on here. AFAICS there are no CDC mutations yet, they are created by augment_mutation_call. Then they are erased for whatever reason (why erase the augmentations, isn't it easier to move out the original mutation?). Then the code duplicates mutate_internal branch, thus any time we will need to extend this logic, we'll need to do it twice.

Reply:
"Then they are erased for whatever reason (why erase the augmentations, isn't it easier to move out the original mutation?)"
augment_mutation_call takes vector of mutations and returns vector of mutations that includes all initial mutations + cdc log mutations for some of those initial mutation. The one that have CDC enabled. Initial mutations are removed because they are paxos mutations and have to be handled differently.

2. "Then the code duplicates mutate_internal branch"
No it does not. Those two calls are different mutate_internal overloads. It's LWT magic so I feel uncomfortable explaining it to you :)

Reply:
I agree it's weird. We could have both code paths (with cdc and without cdc) generate a vector of mutations, and then have just one place where mutate_internal() is called for this order.
Alternatively, if for clarity of this code you switch around the order (this is the option I suggested below) - first write the base write, and only then optionally write the cdc one, then maybe there is some logic in this code (which creates a vector and then removes the original cdc element), because it makes the non-cdc case clearer, with the downside of making the cdc-specific code uglier.

Reply:
using single mutate_internal won't work because those are two different mutate_internal overloads. Could you elaborate a bit more @nyh how you would like to merge those two overloads? The overloads have different logic not just different parameters. overload that takes paxos decision is special.

Reply:
Where do CDC mutations go?

Reply:
Sorry, you're right, the branch is not duplicated, overload of mutate_internal() got me confused.

Reply:
I added a comment WRT that into the code.